README.md
# laravel-itunes-store-api

Here is to have a simple way to interact with the iTunes Store API from a Laravel >= 5.3 app.

This package is usable in production, but should still be considered as a _work in progress_. Found a bug? Got a feature request? Open an issue or better, send a PR!

## Install

You can install this package via composer:

``` bash
$ composer require atomescrochus/laravel-itunes-search-api
```

If you are _on Laravel >=5.5, the package will be working with the auto-discovery feature_. For earlier versions, you will have to install the package' service provider and alias:

```php
// config/app.php
'providers' => [
    ...
    Atomescrochus\ItunesStore\ItunesSearchAPIProvider::class,
];

// no need to add aliases, the packages bind itselfs as "ItunesSearch"
```

You will have to publish the configuration files also if you want to change the default value:
```bash
php artisan vendor:publish --provider="Atomescrochus\ItunesStore\ItunesSearchAPIProvider" --tag="config"
```

## Usage

``` php
// here is an example query to search iTunes Store's API

// Set cache duration as an integer (number of minutes), can be 0
ItunesSearch::setCacheDuration(120) // optional, default is set in config

// You can execute a basic search, and hope for the best
$results = ItunesSearch::query("poker face lady gaga"); // limited to 15 results by default

// You can also send an optional array of other parameters supported by the API, for example
$results = ItunesSearch::query("poker face lady gaga", ['country' => 'CA', 'limit' => 10]);

// You can also execute a lookup (https://affiliate.itunes.apple.com/resources/documentation/itunes-store-web-service-search-api/#lookup)
$results = ItunesSearch::lookup(902122445); // defaults to lookup by iTunes Store ID
$results = ItunesSearch::lookup(468749, 'amgArtistId') // you can do other type of lookups
$results = ItunesSearch::lookup("468749,909253", 'amgArtistId'); // you can also search for multiple ids like so

// And like the query() method, you can send an optional array of parameters
$results = ItunesSearch::lookup(909253, 'id', ['country' => 'CA', 'limit' => 10]); 

// If you ever find yourself that you want to check your local cache for a result, without actually
// polling the Search API (for example, in cache you already know you've been rate limited):
$results = ItunesSearch::cacheOnly()->query("poker face lady gaga");
// if you want to poll again after setting cacheOnly(), you need to reverse the behavior
$results = ItunesSearch::cacheOnly(false)->lookup(909253); 

// In case you only want to check if the query exists in the cache, without actually pulling the data:
$existInCache = ItunesSearch::inCache()->query("poker face lady gaga"); // returns boolean
$existInCache = ItunesSearch::inCache()->lookup(909253); // returns boolean
```

### Caching and iTunes Store API's rate limiting
Curently, the API is "[limited to approximately 20 calls per minute (subject to change)](https://affiliate.itunes.apple.com/resources/documentation/itunes-store-web-service-search-api/)". (Approvimately!)

For now, the only way to know that you're on the rather erratic Store's rate limit is if we hit an HTTP response of `403 Forbidden`. There is no way to know when it expires, or how many call you have left, or anything usefull for that matter (yep, this sucks).

To help you manage the rate limiting, we  provide a parameter to the result object returned by the search called `rateLimited`.  If set to `true`, we encountered a `403` and it means that you are rate limited.

Of course, we cannot stop you from hitting the API even if you are rate limited, so it's _your_ duty to make sure you stope for a little while if `rateLimited` is set to true.

One last thing on rate limiting: since we usually cache results by default, if we ever encounter a `403`, we return an empty result _without_ caching results, without consideration to the caching setting you could have set. This way, if you make the same call again within the normal caching time, but are not rate limited again, you won't get an empty result.

### Results
 
In the example above, what is returned in `$results` is an object containing: a collection of results; a count value for the results; a boolean to know if we are rate limited, a boolean to know if the result is coming from cached data, a boolean to know if we requested to search in the cache only, raw response data; and the unformatted query sent to the API.

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

Soon.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Security

If you discover any security related issues, please email jp@atomescroch.us instead of using the issue tracker.

## Credits

- [Jean-Philippe Murray][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

CHANGELOG.md
# Changelog

All Notable changes to `laravel-itunes-store-api` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## 2.3.0 - 2017-09-03
### Added
- Package auto-discovery for Laravel 5.5

## 2.2.0 - 2016-10-13
### Added
- Method `cacheOnly()` to try and search only in the local cache, without polling the Search API
- Method `inCache()` to check if a request is included in the local cache

## 2.1.2 - 2016-10-13
### Fixed
- Stupid mistakes made on the thrill of the moment

## 2.1.1 - 2016-10-13
### Fixed
- A default response object is returned when there is no result body

## 2.1.0 - 2016-10-13
### Added
- `lookup()` method to search the store by IDs
- Throws an exception for invalid lookup types
- Throws an exception for invalid parameter

## 2.0.1 - 2016-10-13
### Fixed
- Caching logic

## 2.0.0 - 2016-10-13
### Added
- Now with facades, as `ItunesSearch`

### Fixed
- Setting a cache time of `0` now corrently ignore caching

### Deprecated
- Method `search()` is now `query()` to be more consistent with actual usage
- Method `executeSearch()` is now `search()` to be more consistent with actual usage
- Method `cache()` is now `setCacheDuration()` to be more consistent with actual usage

## 1.0.3 - 2016-10-13
### Fixed
- Cache name would change on every request, causing the package to store, but not being able to retreive old cached items

## 1.0.2 - 2016-10-12
### Added
- Fixed a typo in composer.json making the pakcage unrocognizable in packagist

## 1.0.1 - 2016-10-12
### Added
- A parameter to the result object to know if the result is coming from cached results

## 1.0.0 - 2016-10-12
### Added
- Config file with default cache time
- Added a search method to search the Store API
- Added a cache method to set a number of minutes to cache for, overriding the default one in config
- Added a mechanism to guess if we're rate limited and a parameter to the result object to display if we are

## 0.0.1 - 2016-10-12

### Added
- Base config. Tagging a release for packagist convenience

## Please ignore below this line!
## Unreleased - YYYY-MM-DD
### Added
### Deprecated
### Fixed
### Removed
### Security


src/ItunesSearchAPIProvider.php
<?php

namespace Atomescrochus\ItunesStore;

use Illuminate\Support\ServiceProvider;

class ItunesSearchAPIProvider extends ServiceProvider
{
    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot()
    {
        $this->publishes([
            __DIR__.'/config/laravel-itunes-search-api.php' => config_path('laravel-itunes-search-api.php'),
        ], 'config');

        $this->mergeConfigFrom(__DIR__.'/config/laravel-itunes-search-api.php', 'laravel-itunes-search-api');
    }

    /**
     * Register any package services.
     *
     * @return void
     */
    public function register()
    {
        $this->app->bind('atomescrochus.itunes', function () {
            return new ItunesSearchAPI();
        });

        $this->app->booting(function () {
            $loader = \Illuminate\Foundation\AliasLoader::getInstance();
            $loader->alias('ItunesSearch', 'Atomescrochus\ItunesStore\Facades\ItunesSearch');
        });
    }
}


src/ItunesSearchAPI.php
<?php

namespace Atomescrochus\ItunesStore;

use Atomescrochus\ItunesStore\Exceptions\UsageErrors;
use Illuminate\Support\Facades\Cache;

class ItunesSearchAPI
{
    private $cache_time;
    private $searchApiRateLimit;
    private $request_url;
    private $parameters;
    private $endpoint;
    private $possible_lookup_types;
    private $possible_parameters;
    private $cacheOnly;
    private $inCache;

    public function __construct()
    {
        $this->cache_time = empty(config('laravel-itunes-search-api.cache')) ? 60 : config('laravel-itunes-search-api.cache');
        $this->searchApiRateLimit = (object) ['numberOfCalls' => 20, 'perAmountOfSeconds' => 60];
        $this->request_url = "https://itunes.apple.com";
        $this->parameters = [];
        $this->cacheOnly = false;
        $this->inCache = false;
        $this->possible_lookup_types = [
            'id',
            'amgArtistId',
            'amgAlbumId',
            'amgVideoId',
            'upc',
            'isbn'
        ];
        $this->possible_parameters = [
            'term',
            'country',
            'media',
            'entity',
            'attribute',
            'callback',
            'limit',
            'lang',
            'version',
            'explicit'

        ];
    }

    public function setCacheDuration(int $minutes)
    {
        $this->cache_time = $minutes;

        return $this;
    }

    public function cacheOnly(bool $cacheOnly = true)
    {
        $this->cacheOnly = $cacheOnly;

        return $this;
    }

    public function inCache(bool $inCache = true)
    {
        $this->inCache = $inCache;

        return $this;
    }

    public function query($terms, $extra_parameters = ['limit' => 15])
    {
        $this->endpoint = "/search";

        $parameters = array_merge(['term' => $terms], $extra_parameters);
        $this->setParameters($parameters);

        return $this->search();
    }

    public function lookup($id, $type = 'id', $extra_parameters = [])
    {
        if (!in_array($type, $this->possible_lookup_types)) {
            throw UsageErrors::lookupTypes();
        }

        $this->endpoint = "/lookup";
        $parameters = array_merge([$type => $id], $extra_parameters);
        $this->setParameters($parameters, true);

        return $this->search();
    }

    public function search()
    {
        $cache_name = md5($this->getRequestUrl());

        if ($this->inCache) {
            return Cache::has($cache_name);
        }

        $cache = $this->checkForCache($cache_name);
        
        if (isset($cache->content)) {
            $cached = $cache->content;
            $cached->cacheOnly = $this->cacheOnly;
            return $cached;
        }

        if ($this->cacheOnly) { // here, there is no cached content, and we've asked only for the cache...
            return (object) [
                'results' => collect([]),
                'count' => 0,
                'rateLimited' => false,
                'cached' => false,
                'cacheOnly' => true,
                'query' => urldecode($this->getRequestUrl())
            ];
        }

        $response = \Httpful\Request::get($this->getRequestUrl())->expectsJson()->send();
        
        if ($response->code == 200 && $cache->shouldCache == true) {
            return  Cache::remember($cache_name, $this->cache_time, function () use ($response) {
                return $this->formatApiResults($response);
            });
        }

        if ($response->code == 403) {
            return  $this->formatApiResults($response, false, true);
        }

        return  $this->formatApiResults($response, false);
    }

    private function setParameters($parameters, $isLookup = false)
    {
        $possibilities = $this->possible_parameters;

        if ($isLookup == true) {
            $possibilities = array_merge($possibilities, $this->possible_lookup_types);
        }

        foreach ($parameters as $parameter => $value) {
            if (!in_array($parameter, $possibilities)) {
                throw UsageErrors::parameters($parameter);
            }
        }

        $this->parameters = $parameters;
    }

    private function checkForCache($name)
    {
        $cache = (object) ['content' => null, 'shouldCache' => true];

        if (Cache::has($name)) {
            $cache->content = Cache::get($name);
        }

        if ($this->cache_time == 0) {
            $cache->shouldCache = false;
            $cache->content = null;
        }

        return $cache;
    }

    private function formatApiResults($result, $cached = true, $rateLimited = false)
    {
        $raw = $result->raw_body;

        if (isset($result->body)) {
            return (object) [
                'results' => collect($result->body->results),
                'count' => $result->body->resultCount,
                'rateLimited' => $rateLimited,
                'cached' => $cached,
                'cacheOnly' => $this->cacheOnly,
                'raw' => json_decode($raw),
                'query' => urldecode($this->getRequestUrl()),
            ];
        }

        return (object) [
            'results' => collect([]),
            'count' => 0,
            'rateLimited' => $rateLimited,
            'cached' => $cached,
            'raw' => json_decode($raw),
            'query' => urldecode($this->getRequestUrl()),
        ];
    }

    private function getRequestUrl()
    {
        $parameters = http_build_query($this->parameters);
        return "{$this->request_url}{$this->endpoint}?{$parameters}";
    }
}
